packages/tabsheet/src/vaadin-tabsheet.js

```diff
@@ -0,0 +1,278 @@
+ *
+ * @extends HTMLElement
+ * @mixes ElementMixin
+ * @mixes ThemableMixin
+ * @mixes ControllerMixin
+ * @mises DelegateStateMixin
+ */
+class TabSheet extends ControllerMixin(DelegateStateMixin(ElementMixin(ThemableMixin(PolymerElement)))) {
+  static get template() {
+    return html`
+      <style>
+        :host([hidden]) {
+          display: none !important;
+        }
+
+        :host {
+          display: flex;
+          flex-direction: column;
+        }
+
+        [part='tabs-container'] {
+          position: relative;
+          display: flex;
+          align-items: center;
+        }
+
+        ::slotted([slot='tabs']) {
+          flex: 1;
+          align-self: stretch;
+          min-width: 8em;
+        }
+
+        [part='content'] {
+          position: relative;
+          flex: 1;
+          box-sizing: border-box;
+        }
+      </style>
+
+      <div part="tabs-container">
+        <slot name="prefix"></slot>
+        <slot name="tabs"></slot>
+        <slot name="suffix"></slot>
+      </div>
+
+      <vaadin-tabsheet-scroller part="content">
+        <div part="loader"></div>
+        <slot id="panel-slot"></slot>
+      </vaadin-tabsheet-scroller>
+    `;
+  }
+
+  static get is() {
+    return 'vaadin-tabsheet';
+  }
+
+  static get properties() {
+    return {
+      /**
+       * The list of `<vaadin-tab>`s from which a selection can be made.
+       * It is populated from the elements passed inside the slotted
+       * `<vaadin-tabs>`, and updated dynamically when adding or removing items.
+       *
+       * Note: unlike `<vaadin-combo-box>`, this property is read-only.
+       * @type {!Array<!Tab> | undefined}
+       */
+      items: {
+        type: Array,
+        readOnly: true,
+        notify: true,
+      },
+
+      /**
+       * The index of the selected tab.
+       */
+      selected: {
+        value: 0,
+        type: Number,
+        reflectToAttribute: true,
+        notify: true,
+      },
+
+      /**
+       * The slotted <vaadin-tabs> element.
+       */
+      __tabs: {
+        type: Object,
+      },
+
+      /**
+       * The panel elements.
+       */
+      __panels: {
+        type: Array,
+      },
+    };
+  }
+
+  /** @override */
+  static get delegateProps() {
+    return ['selected'];
+  }
+
+  /** @override */
+  static get delegateAttrs() {
+    return ['theme'];
+  }
+
+  /** @protected */
+  ready() {
+    super.ready();
+
+    const tabsItemsChangedListener = () => this._setItems(this.__tabs.items);
+
+    const tabsSelectedChangedListener = () => {
+      this.selected = this.__tabs.selected;
+    };
+
+    this.__overflowController = new OverflowController(this, this.shadowRoot.querySelector('[part="content"]'));
+    this.addController(this.__overflowController);
+
+    // Observe the tabs slot for a <vaadin-tabs> element.
+    this._tabsSlotController = new (class extends SlotController {
```

Comment:
It would be good to move controller class out of the method. Not a blocker, can be done separately.

Reply:
Done